SymbolicT experiment

bts opened the pull request:
Hi Levent,

(/cc @joelburget)

Related to #391 I spent some time experimenting with adding `SymbolicT` to the codebase (I have not yet tried adding `QueryT`).

For this experiment, the crux of the changeset is changing:

```haskell
newtype Symbolic a = Symbolic (ReaderT State IO a)
```
to
```haskell
newtype SymbolicT m a = SymbolicT (ReaderT State m a)
type Symbolic = SymbolicT IO
```

and as a result, functions like
```haskell
runSymbolic :: SBVRunMode -> Symbolic a -> IO (a, Result)
```
to
```haskell
runSymbolic :: MonadIO m => SBVRunMode -> SymbolicT m a -> m (a, Result)
```

Positive consequence:
- clients can now use `SymbolicT` when `Symbolic` doesn't suffice (e.g. when they want to use `Symbolic`, but not strictly as the _base_ of their monad stack, like in https://github.com/LeventErkok/sbv/issues/329#issuecomment-326247068)

Neutral/slightly negative consequence:
- had to add `MonadIO` all over the place

Negative consequences of this approach that we can probably improve upon:
- had to add type annotations to multiple examples, to specify the `m` that `SymbolicT` is using (e.g. see https://github.com/bts/sbv/blob/8be953216e2f65d11490f47ca536ac644be35c0d/Documentation/SBV/Examples/Existentials/CRCPolynomial.hs#L61)
- users of `Symbolic` as currently defined really have to know about the fact that it's defined as `SymbolicT IO` when dealing with type errors

I say that we can "probably improve" because I imagine that we could so something like exposing all of this generalized `SymbolicT`-based machinery from something like `Data.SBV.Trans`, and expose monomorphized (e.g. `Symbolic`- and `IO`-centric, recalling that `type Symbolic = SymbolicT IO`) versions of these functions from `Data.SBV`. That is, people could opt into the generalized machinery only if they need to.

Negative consequences that I'm not sure we can improve upon:
- had to add an extra (`m`) type parameter to both `Provable` and `SExecutable` that I don't know is possible to remove. This seemed necessary in order to write instances like `instance MonadIO m => Provable m (SymbolicT m ())` and `instance (MonadIO m, NFData a) => SExecutable m (SymbolicT m a)`, to get typechecking working where we'd now have (e.g. for the latter) `sName_ :: a -> SymbolicT m ()` concretized to `sName_ :: SymbolicT m1 -> SymbolicT m2 ()` and we want to be able to unify `m1` and `m2`. Adding this extra typeclass parameter was _not_ necessary for e.g. `Metric` because we never define an instance for symbolic computations. This extra type parameter is my biggest gripe with this changeset -- maybe there is something else we could do?
- had to use `FlexibleContexts` in one example instance of client code (`Documentation/SBV/Examples/BitPrecise/BitTricks.hs`)

Note that in this changeset I did _not_ add something like `MonadSymbolic` which might abstract over _any_ monad stack that includes `SymbolicT`. If we had something like this, functions like `forall :: MonadIO m => String -> SymbolicT m (SBV a)` might instead have a signature like `forall :: MonadSymbolic m => String -> m (SBV a)`, allowing the return value to be injected into *any* stack, not just one with `SymbolicT` at the outermost layer. I experimented with something like this as well, but started running into issues and complexity with `Provable` and `SExecutable` so I decided to stop at this point and report back to you.

Brian

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/leventerkok/sbv/431)
<!-- Reviewable:end -->

LeventErkok commented:
@bts Thanks.. Some of the complexity around `SExecutable` and `Provable` are unfortunate, but not unexpected I suppose.

I'll wait till you give me the green light that this branch is ready to merge; I expect you want to do some extra experiments to make sure it fits with your code base first?

bts commented:
@LeventErkok we don't need `ProvableT` and `SExecutableT` because `Provable` and `SExecutable` are typeclasses and not concrete monadic type constructors like `Symbolic` was before the changeset. We _could_ try to add `QueryT`, though (which Joel and I have not yet needed, but I imagine it could probably be done similarly to the changes in this PR for `Symbolic` -> `SymbolicT`)

If I understand correctly, `type Predicate = Symbolic SBool` could make some of the annotations a little more succinct, but I think I am going to try something like this `Data.SBV.Trans` approach, which would theoretically eliminate these annotations from the example code (and, naturally, in the code of people already using sbv).

I will take a look at this and report back.

LeventErkok commented:
Ah, I like the `Data.SBV.Trans` idea. If this can be contained and exposed only to people who care, that'd be ideal.

LeventErkok commented:
Sounds good to me. I think we should have `Provable` not have the `m` argument. I recall I previously suggested `ProvableT` but that doesn't make sense of course. Let's call it `MProvable` and do the simplification; that way there's no cost to users at all.

Please rebase once you're happy with how it works in the Pact sourcebase. Travis builds are still failing though, Did you forget to check something in?

bts commented:
Will do.

Re: travis failures, I see that I've been working in `ghcid` and haven't gotten errors that have cropped up in `SBVTestSuite` during my changes. I will fix this.

LeventErkok commented:
Ah, I was bitten by the same before. I think you can tell ghcid to load both Data/SBV.hs and something from the testsuite to avoid that gotcha

LeventErkok commented:
@bts Is this ready to merge, or are you still working on it?

Also, would be great to rebase on master!

bts commented:
I still need to:
- [ ] finish testing this in pact
- [ ] get rid of the merge commit
- [ ] probably add `QueryT` as well, while I'm at it. hopefully this is quick

LeventErkok commented:
Cool! Thanks for all the great work on this.

bts commented:
@LeventErkok I noticed that it seems that `Symbolic` _could_ be implemented purely -- would you be adverse to trying `SymbolicT` implemented via `StateT` instead of `ReaderT`+`IO`? Then many uses of `IO`+`MonadIO` would drop away. If you would like I could benchmark the results of this.

It seems that we could likewise drop `IO`/`MonadIO` requirements from some `Query` operations as well, where it's possible to use pure state instead of using `IO` to modify the `State`

bts commented:
I guess the other requirement to make `Symbolic` pure would be to explicitly thread `StdGen` to `randomCW` instead of using `randomRIO` as it does now.

A potential benefit of this could be better reproducibility of issues using sbv -- upon an error we could print the initial seed that `Symbolic` was provided.

bts commented:
Ah I completely missed the use of `StableName` -- sorry about that! Thanks!

bts commented:
We're getting close.

We should keep in mind that, as it stands, the changeset re-exports some typeclass methods as monomorphized functions. So if a user wants to provide an instance for e.g. `SymWord` or `SExecutable`, they need to import `Data.SBV.Trans`. I could imagine this being confusing to some end users. Perhaps we might want to consider whether we do this monomorphization for typeclass methods at all. (We could still monomorphize functions that aren't typeclass methods.)

Relatedly I have moved methods off of `SymWord` that are identical (i.e. never overridden) from instance to instance, like `forall`, `forall_`, `exists`, `free`, etc. There will be some slight impact here for users that have explicit imports bringing in these functions via `SymWord` -- e.g. `import Data.SBV (SymWord(..))`. If you'd like me to change this back, I can do that.

Also we've changed the name of `Provable` to `MProvable`, which could affect some users (e.g. Pact had an instance of `Provable` for a custom wrapper around `SBV`).

If you want to see the transformer stuff in action, see `Documentation.SBV.Examples.Queries.Transformer`. Note that in addition to using `ExceptT` within `SymbolicT` and `QueryT`, we are able to use `free` and `checkSat` directly in our custom monad stacks due to `MonadSymbolic` and `MonadQuery`.

With all of the changes, I only had to add `{-# LANGUAGE FlexibleContexts #-}` to one of the examples in `Documentation/` (`BitTricks`).

I will add more notes here as they occur to me.

LeventErkok commented:
Cool. Glad this is coming together. Regarding: 

> Relatedly I have moved methods off of SymWord that are identical (i.e. never overridden) from instance to instance, like forall, forall_, exists, free, etc. 

I remember having this before; it used to be outside and then I moved them to the class. I think them being in the class gives a cleaner interface, so unless there's a good reason why they should be moved out, keeping them in would be better I think.

bts commented:
More notes:

- I had to introduce this notion of `ExtractIO` to restrict the types of transformers we can place underneath a `QueryT`, given that we need to reduce a transformer stack down to `IO` in `executeQuery`. This basically boils down to the fact that things like `ExceptT` and `WriterT` can live underneath `QueryT`, but things like `ReaderT` can't (because you need to provide the input to a reader to be able to realize an `IO` beneath it).

- The monomorphized functions currently do not have documentation on them. Unfortunately I'm not sure what to do about this. Do we duplicate docs across files (across generalized and monomorphized functions)?

LeventErkok commented:
Re:

> The monomorphized functions currently do not have documentation on them. Unfortunately I'm not sure what to do about this. Do we duplicate docs across files (across generalized and monomorphized functions)?

Can we do something like:

```haskell
- | foo, specialized to IO monad. See 'Trans.foo'
foo :: Some-Mono-Type
foo = Trans.foo
```

With the more general function properly linked in haddock? (I think the above should work with Haddock, but I'm not 100% sure.

bts commented:
I wonder whether we should almost do the opposite -- putting the full docs on the monomorphized functions, so that the docs are more useful by default. This way a user doesn't need to go through one of these links to see what each of the functions do.

bts commented:
And the power users who need the full generalization can take this hit of the link-indirection tax for reading docs.

LeventErkok commented:
Yes; I like that better! It can go both ways:

Generalized one:

```
-- | foo: Generalization of `Mono.foo`
```

Special one:
```
-- | foo: Original description.
--
-- NB. For a version that generalizes over the underlying monad, see `Trans.foo`
```

I was wondering if haddock has a meta-data tag for this sort of thing (like `@since`) but I don't think there's anything special.

bts commented:
I'm almost done with the doc changes, but another note occurred to me. If you look at `MonadSymbolic` and `MonadQuery`, they are basically identical. I understand that there are invariants that keep us from merging the two (i.e. we need to keep `Symbolic` and `Query` as separate steps) but this perhaps this could help us with future refactorings.

Also note that I switched the concrete implementation of `QueryT` to use the `Reader`+`IO` pattern just like `SymbolicT` -- it turns out that we weren't using any functionality from `MonadState` except for `get`.

LeventErkok commented:
`Symbolic`/`Query` distinction is mostly historical actually; if I was designing SBV from scratch, it would only have `Query`. (Which I guess would've been called `Symbolic`.) It's a different style of programming, one that has a `sat` view mostly, but it is easier to understand since it matches with how SMT solvers work in general. In a (currently hypothetical) SBV 2.0 rewrite, I think query mode would be the only thing we'd have; but of course needs to be experimented with.

bts commented:
@LeventErkok How do you think the docs look? Do you have any other concerns about the PR, or anything I can tweak from here? Thanks!

LeventErkok commented:
I think you've done the due diligence (many thanks!) Documentation is always work in progress; so I'm not worried for a first shot.

Once you're comfortable with everything (i.e., it *works* for you and the test suite is clean), I'll merge this in a branch and do some extra testing. (I've got some other projects using SBV, want to make sure they all are working fine.) Then I'll merge to master and release.

One thing that'd be nice would be to add dedicated tests in the test-suite to make sure basic transformer related coding works as expected. There are many examples of this in the test suite, for instance, take a look at: `SBVTestSuite/TestSuite/Queries`

It'd be great if you can create a `SBVTestSuite/TestSuite/Trans` directory (or call it whatever you like) and start adding some tests there using that infrastructure. The idea is not to have everything comprehensively tested, but at least have the basics covered and grow the test suite as needed over time.

bts commented:
Great. I've loaded the changes in the Pact codebase and the test suite is passing there. At this point it's probably ready for you to test.

I've added a transformer example to `Documentation`, but I'll add some tests to `SBVTestSuite` as well and report back.

Thanks!

LeventErkok commented:
Cool; let me know when the tests are all in and it's good to go. Can you also rebase the whole thing on top of master first?

bts commented:
@LeventErkok I've added tests to accompany my symbolic evaluation example, and the PR is currently rebased off of `master`.

bts commented:
Sure thing. I am traveling around today / this weekend but will try to get to this shortly.
